gramps_webapi/api/resources/filters.py

```diff
@@ -0,0 +1,298 @@
+        missing="and",
+        validate=validate.OneOf(["and", "or", "xor", "one"]),
+    )
+    i = fields.Boolean(required=False, missing=False)
+    r = fields.List(fields.Nested(QueryRuleSchema))
+
+
+class FilterResource(ProtectedResource, GrampsJSONEncoder):
+    """Filter resource."""
+
+    @use_args(
+        {"filter": fields.Str(), "rule": fields.Str()},
+        location="query",
+    )
+    def get(self, args: Dict, namespace: str) -> Response:
+        """Get available custom filters and rules."""
+        try:
+            namespace = GRAMPS_NAMESPACES[namespace]
+        except KeyError:
+            abort(400)
+
+        rule_list = []
+        for rule_class in _RULES_LOOKUP[namespace]:
+            add_rule = True
+            if args.get("rule") and args["rule"] != rule_class.__name__:
+                add_rule = False
+            if add_rule:
+                rule_list.append(
+                    {
+                        "category": rule_class.category,
+                        "description": rule_class.description,
+                        "labels": rule_class.labels,
+                        "name": rule_class.name,
+                        "rule": rule_class.__name__,
+                    }
+                )
+        if args.get("rule") and not args.get("filter"):
+            return self.response(200, {"rules": rule_list})
+        filter_list = []
+        filters.reload_custom_filters()
+        for filter_class in filters.CustomFilters.get_filters(namespace):
+            add_filter = True
+            if args.get("filter") and args["filter"] != filter_class.get_name():
+                add_filter = False
+            if add_filter:
+                filter_list.append(
+                    {
+                        "comment": filter_class.get_comment(),
+                        "function": filter_class.get_logical_op(),
+                        "invert": filter_class.invert,
+                        "name": filter_class.get_name(),
+                        "rules": [
+                            {
+                                "name": filter_rule.__class__.__name__,
+                                "regex": filter_rule.use_regex,
+                                "values": filter_rule.values(),
+                            }
+                            for filter_rule in filter_class.get_rules()
+                        ],
+                    }
+                )
+        if args.get("filter") and not args.get("rule"):
+            return self.response(200, {"filters": filter_list})
+        return self.response(200, {"filters": filter_list, "rules": rule_list})
+
+    @use_args(FilterSchema(), location="json")
+    def post(self, args: Dict, namespace: str) -> Response:
+        """Create a custom filter."""
+        try:
+            namespace = GRAMPS_NAMESPACES[namespace]
+        except KeyError:
+            abort(404)
+
+        new_filter = build_filter(args, namespace, keys="long")
```

Comment:
A function argument that only has two valid values (in this case "long" and "short") cries out being a boolean rather than a string where all except two values are undefined. `long_keys=True`.

Reply:
Yes, good point. That will go away with a refactor.